### PR TITLE
[Merged by Bors] - Druid DB Connection

### PR DESCRIPTION
**Pull request by fhennig**
## Description

This PR refactors the init process and adds a new custom resource for druid connections.

The Init resource is now called SupersetDB and also tracks a status. it is first `Provisioned`. If possible (secret also deployed) a kubernetes `Job` is started and the status of the CR becomes `Initializing`. The job is watched, and depending on whether it successfully completes or not, the SupersetDB also becomes `Ready` or `Failed` respectively.

This change allows us to also depend on the DB being ready (or not). This is necessary for writing the druid connection information into the DB.

A new CRD is introduced: `DruidConnection`. It holds information about the superset DB it should write to, and the Druid cluster that should be connected. A new controller watches the discovery config map as well as the superset DB resources. Once all the dependencies are there (and `Ready`) a job is started, and the status is updated, just like it is done for the `SupersetDB`.

This is in line with the declarative philosophy of k8s; the spec reflects the desired state and the status tracks the actual underlying cluster. Also, with multiple controllers watching object states, control flow is delegated a bit to the API server. Before we were watching the job state manually, now we react to status changes which are given to us by the API server. The controllers only react to object changes.

This PR also relies on the Druid Discovery PR https://github.com/stackabletech/druid-operator/pull/102

What wasn't done: If the DruidConnection resource is deleted, the actual connection is not deleted. The issue doesn't say that this is necessary, and there is no obvious, documented way to delete data sources from the commandline, we'd have to go into the database and delete stuff there. On the controller side we can use a [finalizer](https://github.com/kube-rs/kube-rs/blob/d90ee9706923ceddf94c3d50928bc390543af2fc/examples/secret_syncer.rs#L92-L102).

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by fhennig** [comment]
I got some feedback from Sönke that the DruidConnection should not reference the DB object, instead it should reference the superset cluster, that makes more sense for the user

File: rust/crd/src/supersetdb.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                .with_recommended_labels(superset, APP_NAME, version, "", "") // TODO fill in missing fields
+                .build(),
+            spec: SupersetDBSpec {
+                superset_version: version.to_string(),
+                credentials_secret: superset.spec.credentials_secret.clone(),
+                load_examples: superset.spec.load_examples_on_init.unwrap_or_default(),
+            },
+            status: Some(SupersetDBStatus::new()),
+        })
+    }
```
Comment (maltesander) [inline rust/crd/src/supersetdb.rs:66]: The rusty way would be to implement `From<SupersetCluster>` (or just into).

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                        .get::<SupersetDB>(
+                            &druid_connection.spec.superset_cluster_name,
+                            Some(&druid_connection.spec.superset_cluster_namespace),
+                        )
+                        .await
+                        .context(SupersetDBRetrieval)?;
+                    // Everything is there, retrieve all necessary info and start the job
+                    let sqlalchemy_str = get_sqlalchemy_uri_for_druid_cluster(
+                        &druid_connection.spec.druid_cluster_name,
+                        &druid_connection.spec.druid_cluster_namespace,
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:157]: These are reoccuring 5 times. I would extract that into an extra variable or offer a method for that within `DruidConnection`

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                        .context(SupersetDBRetrieval)?
+                        .status;
+                    if let Some(s) = superset_db_status {
+                        s.condition == SupersetDBStatusCondition::Ready
+                    } else {
+                        false
+                    }
+                } else {
+                    false
+                };
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:136]: This is a little messy. You get an error (and reconcile) if it does not exist. You have the same behavior from just using get? Maybe something like that
``` 
let mut superset_db_ready = false;
if let Some(status) =
client
    .get::<SupersetDB>(
        &druid_connection.spec.superset_cluster_name,
        Some(&druid_connection.spec.superset_cluster_namespace),
    )
    .await
    .context(SupersetDBRetrieval)?.status {
    superset_db_ready = status.condition == SupersetDBStatusCondition::Ready;
}
```

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    #[snafu(display("could not find {}", superset))]
+    FindSuperset {
+        source: stackable_operator::error::Error,
+        superset: ObjectRef<SupersetCluster>,
+    },
+    #[snafu(display("failed to find superset with name {:?} in namespace {:?}", cluster_ref.name, cluster_ref.namespace))]
+    SupersetClusterNotFound {
+        source: crate::util::Error,
+        cluster_ref: SupersetClusterRef,
+    },
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:46]: Not used (here and in superset db controller)

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+                        .context(ApplyJob)?;
+                    // The job is started, update status to reflect new state
+                    client
+                        .apply_patch_status(FIELD_MANAGER_SCOPE, &druid_connection, &s.importing())
+                        .await
+                        .context(ApplyStatus)?;
+                }
+            }
+            DruidConnectionStatusCondition::Importing => {
+                let ns = druid_connection.metadata.namespace.clone().unwrap();
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:175]: You can call name() and namespace() on the resource (need to import a trait). It should not happen but can you wrap that unwrap in a `NoNamespace` error or something?

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+                let job_name = druid_connection.job_name();
+                let job = client
+                    .get::<Job>(&job_name, Some(&ns))
+                    .await
+                    .context(GetImportJob {
+                        namespace: ns,
+                        name: job_name,
+                    })?;
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:183]: You can use ObjectRef<_>:

```
    #[snafu(display(
        "druid connection state is 'importing' but failed to find job {obj_ref}",
    ))]
    GetImportJob {
        source: stackable_operator::error::Error,
        obj_ref: ObjectRef<Job>,
    },
```
and call
`
.context(GetImportJob {
                       obj_ref: ObjectRef::new(&job_name).within(&ns)
                    })?;
`

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+                if let Some(ns) = new_status {
+                    client
+                        .apply_patch_status(FIELD_MANAGER_SCOPE, &druid_connection, &ns)
+                        .await
+                        .context(ApplyStatus)?;
+                }
+            }
+            DruidConnectionStatusCondition::Ready => (),
+            DruidConnectionStatusCondition::Failed => (),
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:199]: _ => ()

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        .context(GetDruidConnStringConfigMap {
+            cm_name: cluster_name.to_string(),
+            namespace: namespace.to_string(),
+        })?
+        .data
+        .and_then(|mut data| data.remove("DRUID_SQLALCHEMY"))
+        .context(MissingDruidConnString {
+            cm_name: cluster_name.to_string(),
+            namespace: namespace.to_string(),
+        })
```
Comment (maltesander) [inline rust/operator-binary/src/druid_connection_controller.rs:236]: You can use ObjectRef here too.

File: rust/operator-binary/src/main.rs [hunk 9]
```diff
@@ -52,7 +60,7 @@ async fn main() -> anyhow::Result<()> {
         Command::Crd => println!(
             "{}{}",
             serde_yaml::to_string(&SupersetCluster::crd())?,
-            serde_yaml::to_string(&Init::crd())?
+            serde_yaml::to_string(&SupersetDB::crd())?
```
Comment (maltesander) [inline rust/operator-binary/src/main.rs:63]: Druid connection?

File: rust/operator-binary/src/superset_controller.rs [hunk 10]
```diff
@@ -80,25 +92,32 @@ pub async fn reconcile_superset(
...
-        superset_version(&superset)?,
+        superset_version(&superset).context(NoSupersetVersion)?,
         &transform_all_roles_to_config(
             &superset,
             [(
                 SupersetRole::Node.to_string(),
                 (
                     vec![PropertyNameKind::Env],
-                    superset.spec.nodes.clone().context(NoNodeRoleSnafu)?,
+                    superset.spec.nodes.clone().context(NoNodeRole)?,
```
Comment (maltesander) [inline rust/operator-binary/src/superset_controller.rs:110]: Why remove the suffix?

File: rust/operator-binary/src/superset_db_controller.rs [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        source: crate::util::Error,
+        cluster_ref: SupersetClusterRef,
+    },
+    #[snafu(display("object does not refer to SupersetCluster"))]
+    InvalidSupersetReference,
+    #[snafu(display("could not find {}", superset))]
+    FindSuperset {
+        source: stackable_operator::error::Error,
+        superset: ObjectRef<SupersetCluster>,
+    },
```
Comment (maltesander) [inline rust/operator-binary/src/superset_db_controller.rs:47]: Not used

File: rust/operator-binary/src/superset_db_controller.rs [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+                // TODO we need to fetch the job here
+                // we need namespace/name.
+                let ns = superset_db.metadata.namespace.clone().unwrap();
+                let job_name = superset_db.job_name();
+                let job = client.get::<Job>(&job_name, Some(&ns)).await.context(
+                    GetInitializationJob {
+                        namespace: ns,
+                        name: job_name,
+                    },
+                )?;
```
Comment (maltesander) [inline rust/operator-binary/src/superset_db_controller.rs:128]: name(), namespace(), ObjectRef<_> -> check druid_connection_controller

**Review by maltesander — CHANGES_REQUESTED** [review]
Tested and works. Some minor things.

File: Cargo.toml [hunk 13]
```diff
@@ -1,2 +1,6 @@
 [workspace]
 members = ["rust/crd", "rust/operator-binary"]
+
+# TODO remove this again before merging into main
+[patch."git+https://github.com/stackabletech/operator-rs.git"]
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
```
[the remark this answers is not in the document]
Comment (fhennig) [reply Cargo.toml]: I made a new release for operator-rs and refernced the tag now. I've also added it to the changelog

File: rust/operator-binary/src/main.rs (hunk 9 above)
Comment (fhennig) [reply rust/operator-binary/src/main.rs:63]: Added

File: rust/operator-binary/src/superset_controller.rs (hunk 10 above)
Comment (fhennig) [reply rust/operator-binary/src/superset_controller.rs:110]: I was lazy :D I'll add it back

**Comment by fhennig** [comment]
TODO: Move `DruidConnection` and `SupersetDB` into the `superset` apigroup (they are in the `command` api group at the moment)

**Comment by fhennig** [comment]
There's a problem with the init job, it doesn't fail if it can't connect to the database. instead id finishes "successfully" but actually it didn't do its job.

**Comment by fhennig** [comment]
When the cluster controller writes the supersetdb object, it overwrites the status if the object exists already. it shouldn't do that

**Comment by fhennig** [comment]
[Here](https://github.com/stackabletech/superset-operator/blob/d088d32c919a7e3412e71fe953047ded8f389b30/rust/operator-binary/src/superset_db_controller.rs#L165) we should do `bash -euo pipefail -c`

File: rust/crd/src/druidconnection.rs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        k8s_openapi = "stackable_operator::k8s_openapi",
+        schemars = "stackable_operator::schemars"
+    )
+)]
+#[serde(rename_all = "camelCase")]
+pub struct DruidConnectionSpec {
+    pub superset_cluster_name: String,
+    pub superset_cluster_namespace: String,
+    pub druid_cluster_name: String,
+    pub druid_cluster_namespace: String,
```
[the remark this answers is not in the document]
Comment (fhennig) [reply rust/crd/src/druidconnection.rs]: done

File: rust/crd/src/supersetdb.rs (hunk 1 above)
Comment (fhennig) [reply rust/crd/src/supersetdb.rs]: this is just the database to go along with the cluster, `From` and `Into` are conversions which are semantically different, so I wouldn't use the traits here.

File: rust/operator-binary/src/druid_connection_controller.rs (hunk 2 above)
Comment (fhennig) [reply rust/operator-binary/src/druid_connection_controller.rs]: I did my best to use ObjectRefs now

File: rust/operator-binary/src/druid_connection_controller.rs (hunk 8 above)
Comment (fhennig) [reply rust/operator-binary/src/druid_connection_controller.rs]: done

File: rust/operator-binary/src/superset_db_controller.rs (hunk 12 above)
Comment (fhennig) [reply rust/operator-binary/src/superset_db_controller.rs]: I did it

File: rust/operator-binary/src/druid_connection_controller.rs (hunk 3 above)
Comment (fhennig) [reply rust/operator-binary/src/druid_connection_controller.rs]: much better

File: rust/operator-binary/src/superset_controller.rs (hunk 10 above)
Comment (fhennig) [reply rust/operator-binary/src/superset_controller.rs]: done

File: rust/crd/src/druidconnection.rs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+use stackable_operator::k8s_openapi::chrono::Utc;
+use stackable_operator::kube::CustomResource;
+use stackable_operator::kube::ResourceExt;
+use stackable_operator::schemars::{self, JsonSchema};
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ClusterRef {
+    pub name: String,
+    pub namespace: String,
```
Comment (maltesander) [inline rust/crd/src/druidconnection.rs:12]: Im sorry i suggested it is this way. The way to go should be an optional namespace.
(I stumbled over that while adapting KUTTL tests, which run in random namespaces). We should "default" to the superset cluster namespace if the `ClusterRef` namespace is `None`.
You can have a look here: https://github.com/stackabletech/trino-operator/pull/95
I had to adapt a lot in Trino to make it namespace independent.

**Review by maltesander — APPROVED** [review]
LGTM.
I had the one comment for namespaces. This is not part of this PR so im happy to merge as is (and create an issue and (or) PR where we address this).

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by fhennig** [comment]
bors merge

**Pull request closed without merging**